### PR TITLE
reduce SLOAD calls in _depositInVault

### PR DESCRIPTION
**Pull request by kamescg**
https://github.com/code-423n4/2021-06-pooltogether-findings/issues/44

### CODE REVIEW

File: contracts/yield-source/YearnV2YieldSource.sol [hunk 1]
```diff
@@ -168,8 +168,9 @@ contract YearnV2YieldSource is IYieldSource, ERC20Upgradeable, OwnableUpgradeabl
     /// @return The actual amount of shares that were received for the deposited tokens
     function _depositInVault() internal returns (uint256) {
         IYVaultV2 v = vault; // NOTE: for gas usage
-        if(token.allowance(address(this), address(v)) < token.balanceOf(address(this))) {
-            token.safeApprove(address(v), type(uint256).max);
+        IERC20Upgradeable _token = token;
+        if(_token.allowance(address(this), address(v)) < _token.balanceOf(address(this))) {
```
Comment (PierrickGT) [inline contracts/yield-source/YearnV2YieldSource.sol]: Add some space
```suggestion
        IERC20Upgradeable _token = token;
        
        if (_token.allowance(address(this), address(v)) < _token.balanceOf(address(this))) {
```

File: contracts/yield-source/YearnV2YieldSource.sol [hunk 2]
```diff
@@ -168,8 +168,9 @@ contract YearnV2YieldSource is IYieldSource, ERC20Upgradeable, OwnableUpgradeabl
...
+            _token.safeApprove(address(v), type(uint256).max);
```
Comment (PierrickGT) [inline contracts/yield-source/YearnV2YieldSource.sol:173]: We should do this in `initialize`, this way we wouldn't have to write all this `if`.

File: contracts/yield-source/YearnV2YieldSource.sol [hunk 3]
```diff
@@ -181,15 +182,17 @@ contract YearnV2YieldSource is IYieldSource, ERC20Upgradeable, OwnableUpgradeabl
...
     function _withdrawFromVault(uint amount) internal returns (uint256) {
+        IERC20Upgradeable _token = token;
         uint256 yShares = _tokenToYShares(amount);
-        uint256 previousBalance = token.balanceOf(address(this));
+        uint256 previousBalance = _token.balanceOf(address(this));
         // we accept losses to avoid being locked in the Vault (if losses happened for some reason)
-        if(maxLosses != 0) {
-            vault.withdraw(yShares, address(this), maxLosses);
+        uint256 _maxLosses = maxLosses;
+        if(_maxLosses != 0) {
```
Comment (PierrickGT) [inline contracts/yield-source/YearnV2YieldSource.sol]: Add a space
```suggestion
        if (_maxLosses != 0) {
```

**Review by PierrickGT — CHANGES_REQUESTED** [review]

**Comment by PierrickGT** [comment]
As a side note that should be addressed in another PR, we don't need to `initialize` with `token` since we can retrieve the underlying token address with `_vault.token()`.
https://github.com/pooltogether/pooltogether-yearnv2-yield-source/blob/4277e8357fa9b505c3e5ce8e1a3a166078f4b355/contracts/yield-source/YearnV2YieldSource.sol#L68

Which would avoid us doing this check: https://github.com/pooltogether/pooltogether-yearnv2-yield-source/blob/4277e8357fa9b505c3e5ce8e1a3a166078f4b355/contracts/yield-source/YearnV2YieldSource.sol#L74

Fixed in this PR: https://github.com/pooltogether/pooltogether-yearnv2-yield-source/pull/8

**Comment by PierrickGT** [comment]
Not sure if it will lower gas consumption since `token` is an `internal` variable anyway?
https://github.com/pooltogether/pooltogether-yearnv2-yield-source/blob/bff887744d1ef487aebf3be74e1b8070683d945b/contracts/yield-source/YearnV2YieldSource.sol#L27

File: contracts/yield-source/YearnV2YieldSource.sol (hunk 2 above)
Comment (asselstine) [reply contracts/yield-source/YearnV2YieldSource.sol:173]: That's covered here: https://github.com/code-423n4/2021-06-pooltogether-findings/issues/95

**Review by asselstine — APPROVED** [review]

**Review by asselstine — APPROVED** [review]

**Pull request merged**